Add missing_elements method

ineverov opened the pull request:
It returns expected elements which are not present on the page/section

Resolves issue https://github.com/site-prism/site_prism/issues/48

luke-hill commented on features/step_definitions/element_steps.rb:
```diff
@@ -55,16 +55,23 @@
+  mapped_item_names =
+    @test_site.dynamic.class.mapped_items.map(&:values).flatten
+
+  expect(@test_site.dynamic.missing_elements).to match_array(mapped_item_names)
```
this doesn't make sense to me. Because the dynamic page has a bunch of elements that are present. So surely the missing elements on dynamic page should be an empty array? Or am I missing something?

ineverov replied:
I used the same approach as the negative test above (I'd not say it's good) [here](https://github.com/site-prism/site_prism/pull/52/files#diff-1f9996f18a011c66b2e4de840767086eR72) and [here](https://github.com/site-prism/site_prism/pull/52/files#diff-5e497d00aedf9e13bd7fefa6a06e8126R63). 
[Here](https://github.com/site-prism/site_prism/pull/52/files#diff-1f9996f18a011c66b2e4de840767086eR6) in the background we navigate to the home page but checking the dynamic page elements

luke-hill replied:
No I understand the naming, just the outcome seems odd. Dynamic page should have *no* missing elements. So the fact you're testing this page with missing elements feels odd to me.

luke-hill replied:
Rebase with master, and then hopefully I can re-review and we can look into merging this @ineverov

ineverov replied:
@luke-hill rebased and fixed

luke-hill commented on spec/site_prism/element_checker_spec.rb:
```diff
@@ -105,6 +105,19 @@
+
+        missing.each { |name| expect(page).to receive(:there?).with(name).once.and_return(false) }
+
+        expect(page.missing_elements).to eql(missing)
```
Do we need eql here?

ineverov replied:
replaced with `match_array`

luke-hill commented on spec/site_prism/element_checker_spec.rb:
```diff
@@ -105,6 +105,19 @@
+      let(:missing) { expected_items[0..1] }
+
+      it 'returns missing elements' do
+        present.each { |name| expect(page).to receive(:there?).with(name).once.and_call_original }
```
At the moment this won't bork rubocop, but down the line I plan on having 1 spec per it. It's one of the tasks I've got to clean up.

Do you think we "should" have 1 spec per it, and if so, can we refactor this slightly maybe into more tests?

ineverov replied:
I don't see the good way to split test to several its. Potentially replace expect with allow, but in this case, it won't verify it is called for each expected_item. Any suggestions?
I'm usually ok having multiple expects in it if they are closely related to each other.

luke-hill replied:
I don't know either. Happy for you to maybe offer some advice here.

Maybe we could change the logic in the repo to allow 2 assertions. But I would like to try keep as close to rubocop rspec standards as is possible. Could we maybe have multiple tests?

I'm not really sure on this one.

ineverov commented:
@luke-hill I added small README for the new method and renamed the method from `missing_elements` to `elements_missing` as opposite to method `elements_present`

luke-hill commented on README.md:
````diff
@@ -736,6 +736,25 @@ use the bleeding edge version of the logic. Then simply set the following config
 SitePrism.use_all_there_gem = true
 ```
 
+### Getting the list of missing element
````
elements then a new line

luke-hill commented on README.md:
````diff
@@ -736,6 +736,25 @@ use the bleeding edge version of the logic. Then simply set the following config
 ```
 
+### Getting the list of missing element
+If all_there? returns false and you wish to get the list of missing elements for debugging purposes
````
`#all_there?`

luke-hill commented on README.md:
```diff
@@ -736,6 +736,25 @@ use the bleeding edge version of the logic. Then simply set the following config
 
+### Getting the list of missing element
+If all_there? returns false and you wish to get the list of missing elements for debugging purposes
+you may want to use `elements_missing` method. It will return the names of missing elements out of
```
`#elements_missing`

luke-hill commented on README.md:
````diff
@@ -736,6 +736,25 @@ use the bleeding edge version of the logic. Then simply set the following config
+you may want to use `elements_missing` method. It will return the names of missing elements out of
+expected_elements list
+```ruby
+class TestPage < SitePrism::Page
````
Can we use one of the ones above, or something that doesn't contain the word "Page", as it's something I think some people like but I personally don't. I've tried to remove traces of `SomethingSection` and `SomethingPage` from the readme

luke-hill commented on features/page.feature:
```diff
@@ -70,3 +70,7 @@ Feature: Page Properties
+
+  Scenario: Missing Elements
+    When I navigate to a page with no title
+    Then missing elements are returned
```
lets say "all missing elements" to keep consistent with the test on 66

ineverov commented:
@luke-hill I updated specs and README

luke-hill commented on README.md:
```diff
@@ -736,6 +736,26 @@ use the bleeding edge version of the logic. Then simply set the following config
+### Getting the list of missing elements
+
+If `#all_there?` returns false and you wish to get the list of missing elements for debugging purposes
+you may want to use `#elements_missing` method. It will return the names of missing elements out of
```
It will return all missing elements from the `expected_elements` list

luke-hill commented on README.md:
```diff
@@ -736,6 +736,26 @@ use the bleeding edge version of the logic. Then simply set the following config
+If `#all_there?` returns false and you wish to get the list of missing elements for debugging purposes
+you may want to use `#elements_missing` method. It will return the names of missing elements out of
+expected_elements list
+
```
If you do not provide a list of `expected_elements` this method will return **all** elements that are missing on the page; from those which are defined.

luke-hill commented:
Can you rebase, fix final language comments and add a changelog item please Ivan?
